File: androidApp/features/profile/src/main/kotlin/com/velkonost/getbetter/android/features/profile/components/ProfileHeader.kt
package com.velkonost.getbetter.android.features.profile.components

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.velkonost.getbetter.core.compose.components.AppButton
import com.velkonost.getbetter.shared.resources.SharedR
import dev.icerock.moko.resources.compose.colorResource
import dev.icerock.moko.resources.compose.stringResource

@Composable
fun ProfileHeader(
    modifier: Modifier = Modifier,
    isLoading: Boolean = true,
    showSettings: Boolean = true,
    userName: String,
    avatarUrl: String?,
    onAvatarClick: () -> Unit,
    onSettingsClick: () -> Unit,
    isAnonymous: Boolean = false,
    onSignUpClick: (() -> Unit)? = null
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .padding(top = 32.dp)
            .height(128.dp)
    ) {
        Avatar(
            modifier = modifier,
            isLoading = isLoading,
            avatarUrl = avatarUrl,
            onClick = onAvatarClick
        )

        Column {
            if (showSettings && !isAnonymous) {
                Row {
                    Spacer(modifier = modifier.weight(1f))

                    SettingsButton(modifier = modifier) {
                        onSettingsClick.invoke()
                    }
                }
            }

            Spacer(modifier = modifier.weight(1f))

            Row(modifier = modifier.padding(start = 16.dp)) {
                if (!isAnonymous) {
                    Text(
                        text = userName,
                        color = colorResource(resource = SharedR.colors.text_title),
                        style = MaterialTheme.typography.titleLarge
                    )
                } else {
                    AppButton(
                        labelText = stringResource(resource = SharedR.strings.auth_signup_button),
                        isLoading = isLoading,
                        onClick = {
                            onSignUpClick?.invoke()
                        }
                    )
                }
            }
        }
    }
}

File: iosApp/iosApp/Features/Profile/components/ProfileHeader.swift
//
//  ProfileHeader.swift
//  iosApp
//
//  Created by velkonost on 20.09.2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import SharedSDK

struct ProfileHeader: View {
    
    let userName: String
    let avatarUrl: String?
    let isLoading: Bool
    let onAvatarClick: () -> Void
    let onSettingsClick: () -> Void
    let showSettings: Bool
    let isAnonymous: Bool
    let onSignUpClick: (() -> Void)?
    let onBlockUserClick: (() -> Void)?
    
    @State private var confirmBlockDialog: Bool = false
    
    init(
        userName: String, avatarUrl: String?,
        isLoading: Bool,
        showSettings: Bool = true,
        isAnonymous: Bool = false,
        onAvatarClick: @escaping () -> Void,
        onSettingsClick: @escaping () -> Void,
        onSignUpClick: (() -> Void)? = nil,
        onBlockUserClick: (() -> Void)? = nil
    ) {
        self.userName = userName
        self.avatarUrl = avatarUrl
        self.isLoading = isLoading
        self.showSettings = showSettings
        self.isAnonymous = isAnonymous
        
        self.onAvatarClick = onAvatarClick
        self.onSettingsClick = onSettingsClick
        self.onSignUpClick = onSignUpClick
        self.onBlockUserClick = onBlockUserClick
    }
    
    var body: some View {
        HStack(alignment: .bottom) {
            Avatar(isLoading: isLoading, avatarUrl: avatarUrl) {
                onAvatarClick()
            }
            
            VStack {
                if showSettings && !isAnonymous {
                    HStack {
                        Spacer()
                        SettingsButton {
                            onSettingsClick()
                        }
                    }
                } else if !showSettings {
                    BlockUserButton {
                        if onBlockUserClick != nil {
                            confirmBlockDialog = true
                        }
                    }
                }
                
                Spacer()
                
                HStack {
                    if !isAnonymous {
                        Text(userName)
                            .foregroundColor(.textTitle)
                            .style(.titleLarge)
                    } else {
                        AppButton(
                            labelText: SharedR.strings().auth_signup_button.desc().localized(),
                            isLoading: isLoading,
                            onClick: {
                                if onSignUpClick != nil {
                                    onSignUpClick!()
                                }
                            },
                            widthPercent: 0.4
                        )
                    }
                    Spacer()
                }
                .padding(.init(top: .zero, leading: 16, bottom: .zero, trailing: .zero))
            }.frame(height: 128)
            
        }
        .padding(.init(top: 32, leading: .zero, bottom: .zero, trailing: .zero))
        .frame(height: 128)
        .alert(
             SharedR.strings().note_detail_hide_title.desc().localized(), isPresented: $confirmBlockDialog) {
                Button(SharedR.strings().confirm.desc().localized()) {
                    viewModel.dispatch(action: NoteDetailActionHideClick())
                }
                Button(SharedR.strings().cancel.desc().localized(), role: .cancel) {}
            } message: {
                Text(SharedR.strings().note_detail_hide_text.desc().localized())
            }
    }
}
